### PR TITLE
[i3c] Target reset generation

### PR DESCRIPTION
**Pull request by alees24**
_**One module `i3c_target_rst` has been somewhat overlooked in the upstreaming**_

This small module handles the generation of the asynchronous reset into the SCL-clocked Target transceiver logic. It works in conjunction with the `i3c_timers` module to produce a sustained reset assertion in response to:

 - IP block reset.
 - Software-initiated reset.
 - Serious errors reported by the Target-side logic.

It also generates the enable to the Target-side input buffers, after a Bus Available condition is observed following reset. This ensures that the Target transceiver logic is connected to the I3C bus only when it is in a known state, not mid-transfer.

### CODE REVIEW

File: hw/ip/i3c/rtl/i3c_target_rst.sv [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    end else begin
+      reset_trx_q <= resetting;
+      enabled_q   <= enable_i;
+    end
+  end
+
+  // For an individual reset-generating event, this should be a single-cycle pulse; more generally,
+  // the timer (`i3c_timers`) will put the transceiver into reset upon assertion of this output, and
+  // start measuring from its deassertion.
+  assign reset_trx_o = reset_trx_q;
```
Comment (etterli) [inline hw/ip/i3c/rtl/i3c_target_rst.sv]: this does not seem as it is always a pulse?

File: hw/ip/i3c/rtl/i3c_target_rst.sv [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+// Generation of asynchronous reset for the Target-side transceiver logic.
+module i3c_target_rst (
+  // Main IP clock and reset.
+  input  clk_i,
+  input  rst_ni,
+
+  // Control.
+  input  enable_i,
+  input  sw_reset_i,
+  input  te_recov_i,
```
Comment (etterli) [inline hw/ip/i3c/rtl/i3c_target_rst.sv]:
```suggestion
  // Control.
  input  enable_i, // level
  input  sw_reset_i, // pulse
  input  te_recov_i, // pulse
```

File: hw/ip/i3c/rtl/i3c_target_rst.sv [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  // the timer (`i3c_timers`) will put the transceiver into reset upon assertion of this output, and
+  // start measuring from its deassertion.
+  assign reset_trx_o = reset_trx_q;
+
+  // Buffer enable state.
+  logic buf_en_q;
+  always_ff @(posedge clk_i or negedge rst_ni) begin
+    if (!rst_ni) buf_en_q <= 1'b0;  // Input buffers are disabled when the IP block is reset.
+    else if (resetting) buf_en_q <= 1'b0;
+    else if (bus_avail_i) buf_en_q <= enabled_q;
```
Comment (andreaskurth) [inline hw/ip/i3c/rtl/i3c_target_rst.sv:63]: Can `enabled_q` ever be sampled as `0` here? `bus_avail_i` requires the `targ_bus_avail` timer to have reached zero, which requires `targ_bus_active == 0` in the preceding cycle, which in turn requires `targ_enabled == 1` in that cycle (`i3c_core.sv:204`) - so `enabled_q` looks like it is always `1` at this point.

If that's right, the enabled->disabled direction never takes effect: once set, `inbuf_en_o` is only cleared by `resetting` or by IP reset. After software clears `TARG_CONTROL.EN` the SCL/SDA buffers stay connected, the transceiver keeps being clocked by bus traffic and is never put back into reset, and `rstdet_enable_o` stays asserted.

Would adding the disabling condition to `resetting` work - e.g. holding `reset_trx_o` while `!enable_i`, which would also make `enabling` redundant? Or, if remaining connected while disabled is deliberate, would `1'b1` here plus a comment saying why be clearer?

File: hw/ip/i3c/rtl/i3c_target_rst.sv [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+  assign enabling  = enable_i & ~enabled_q;
+  assign resetting = |{enabling, sw_reset_i, te_recov_i};
+
+  always_ff @(posedge clk_i or negedge rst_ni) begin
+    if (!rst_ni) begin
+      // Ensure that the Target transceiver logic is reset whenever the IP block is reset.
+      reset_trx_q <= 1'b1;
+      enabled_q   <= 1'b0;
+    end else begin
+      reset_trx_q <= resetting;
```
Comment (andreaskurth) [inline hw/ip/i3c/rtl/i3c_target_rst.sv:48]: Because `i3c_timers` registers `tm_resets_i`, `targ_trx_rst_n` drops one cycle *after* `inbuf_en_o` clears: `resetting(n)` gives `buf_en_q(n+1) = 0` and `reset_trx_q(n+1) = 1`, but `targ_trx_rst_n(n+2) = 0`.

If `sw_reset_i` asserts while SCL is low, the forced `OutDisabled = 1'b1` produces one spurious posedge into the transceiver in that window, which flips `sda_en_ptog` and clocks `sda_od_en_pq <= sda_od_en_nq` and `sda_pq[0]`. The Target can therefore advance its output early and glitch SDA for up to one IP clock (~20ns at 50MHz) before the reset releases the drivers. Our own state doesn't matter (a 5us reset follows), but another device mid-transfer could sample it. When the bus is idle the gating matches `OutDisabled` and there's no edge at all, so the trigger is narrow.

Probably not worth fixing in this PR; I'm OK with a TODO plus a tracking issue.

File: hw/ip/i3c/rtl/i3c_core.sv [hunk 5]
```diff
@@ -243,6 +243,7 @@ module i3c_core
     .enable_i   (targ_enabled),
     .sw_reset_i (targ_sw_reset),
```
Comment (andreaskurth) [inline hw/ip/i3c/rtl/i3c_core.sv:244]: `targ_sw_reset` is identical to `ctrl_sw_reset`, so `TARG_CONTROL.RESET` never reaches here: `reg2hw.targ_control.reset.q`/`.qe` are generated by `i3c_reg_top` but read nowhere in the design, which makes that register a no-op despite being documented as "writing '1' to this field will reset all of the Target-side logic".

Since `RESET_CONTROL.SOFT_RST` already resets the Target-side logic, this doesn't look critical to me, so a TODO here plus a tracking issue would be fine. If it is wired up, should the term be qualified by `!STBY_CR_SUPPORT`, per the register description?

**Review by andreaskurth — COMMENTED** [review]
Thanks @alees24 for spotting and closing this gap in the upstreaming.

Minor changes requested: two of the reset/disable paths look to me like they don't do what the code reads as doing (both inline), and the remainder are TODO/tracking-issue requests.

I checked that #31034 already covers `targ_trx_rst_n` being generated combinationally after flops, so I haven't raised that here.

Three further items that fall outside the diff hunks, so they don't attach inline:

**`i3c_core.sv:213`** - `targ_status.active` reads `1` about 5us after reset release even while the Target is disabled and its input buffers are disconnected, and it reads `1` after a soft reset while `u_targ_rst` is still waiting for a Bus Available Condition, which on a busy bus can take arbitrarily long. Software therefore has no way to tell when the Target is actually on the bus, since `inbuf_enable` isn't exposed. Given the field is described as "whether the Target functionality is presently active", would ANDing in `inbuf_enable` (or exposing it separately) be more useful? A TODO plus a tracking issue would be fine by me if this isn't needed for minimal functionality.

**Documentation** - the reset/connect sequence isn't described anywhere. `theory_of_operation.md` §"Target-side logic" covers the SCL clocking but not that the Target is deliberately detached from the bus and reconnected only at a Bus Available Condition, and `programmers_guide.md` documents the 5us "Targ Reset" interval in the timer table but not that after `SOFT_RST` the Target stays off the bus for reset + tAVAL *and* until a Bus Available Condition occurs. Could a couple of sentences be added, or a tracking issue filed if the I3C docs are being completed separately?

**`i3c_core.sv:236`** - nit: stray "we" in "when we the bus has been idle". Also, "idle" reads as the IP's separate 200us Bus Idle condition, whereas what is measured here is Bus Available (tAVAL, 1us) - would "available" be more accurate?

File: hw/ip/i3c/rtl/i3c_core.sv (hunk 5 above)
Comment (alees24) [reply hw/ip/i3c/rtl/i3c_core.sv:244]: That was obviously a mistake, one that I found and fixed independently in another PR. #31212 contains the latest 'soft reset' generation, which _may_ be the final version although it's still not yet resolved how these resets are to be handled (#31309).

File: hw/ip/i3c/rtl/i3c_target_rst.sv (hunk 3 above)
Comment (alees24) [reply hw/ip/i3c/rtl/i3c_target_rst.sv]: The action of disabling or resetting had not been checked or tested, to be honest. I've updated the logic with much more attention to such details, and an expectation that the TARGET_CONTROL.RESET mechanism gets implemented as expected in #31309. There is a basic test script that exercises enabling, sw reset and disabling, although it's impractical for this to catch edge cases/races within active traffic.

File: hw/ip/i3c/rtl/i3c_target_rst.sv (hunk 4 above)
Comment (alees24) [reply hw/ip/i3c/rtl/i3c_target_rst.sv:50]: The impact of the reset upon the buffer enables should now be delayed by long enough to avoid this, except in the case of the 'Bus Available' condition when it's harmless anyway. It's worth noting that a randomly-timed reset during transceiver activity should be avoided and the target enable should be deasserted first.

File: hw/ip/i3c/rtl/i3c_core.sv [hunk 6]
```diff
@@ -243,23 +243,32 @@ module i3c_core
...
-    .reset_trx_o(targ_reset_trx),
+    .reset_trig_o(targ_trx_rst_trig),
 
     // Buffer enable for SCL and SDA into the SCL-driven target transceiver logic.
     .inbuf_en_o (inbuf_enable)
   );
 
+  // Ensure that the asynchronous reset into the Target transceiver is glitch free.
+  logic targ_trx_prst_n;
+  prim_flop #(.Width(1), .ResetValue(0)) u_targ_rst_src (
```
Comment (andreaskurth) [inline hw/ip/i3c/rtl/i3c_core.sv:260]: This flop is now the only source of `targ_trx_rst_n`, the asynchronous reset for the whole SCL-clocked domain, but `scan_rst_ni` is still discarded at `i3c_core.sv:144`, so the transceiver can't be reset in scan mode. `spi_device` muxes it into each of its locally generated foreign-domain resets with `prim_clock_mux2 #(.NoFpgaBufG(1'b1))`.

If you'd rather not address this here, #31335 has a DFT bullet, but it names clock gates and RAMs rather than resets.

Separately: does this flop now satisfy #31034? It gives the "reset signal must come from a flop" part, and the release happens with the input buffers already gated off, which looks like the second bullet there. If so, can #31034 be closed or narrowed?

File: hw/ip/i3c/rtl/i3c_core.sv [hunk 7]
```diff
@@ -243,23 +243,32 @@ module i3c_core
...
+    .q_o   (targ_trx_rst_n)
+  );
+
   // Controller-side I3C input buffering.
   logic [NumSDALanes-1:0] ctrl_sda_buf;
   i3c_buf_en #(.Width(NumSDALanes)) u_ctrl_inbufs (
-    // Enable input propagation.
-    // TODO: Probably want to gate off the SDA lane when the bus is available and the Controller
-    //       does not need to drive it. We need a transceiver output to achieve this.
+    // Enable input propagation; disconnect from the SDA input only when the Controller logic;
```
Comment (andreaskurth) [inline hw/ip/i3c/rtl/i3c_core.sv:270]: Two things about this comment.

The first clause looks like it lost a word - "only when the Controller logic" seems to want something like "is disabled".

Is Start Request detection the right justification here, though? `sreq_sda_i` is driven from `ctrl_sda_sync` (line 1449), which bypasses these buffers; `ctrl_sda_buf` only feeds `.sda_i` at line 1452. So the reason to keep this connected whenever the Controller is enabled would be data/ACK/arbitration sampling rather than Start Requests.

This also settles the TODO it replaces, in the opposite direction. Worth capturing that decision somewhere more findable than a comment in this PR?

File: hw/ip/i3c/rtl/i3c_target_rst.sv (hunk 3 above)
Comment (andreaskurth) [reply hw/ip/i3c/rtl/i3c_target_rst.sv]: Thanks, `enabled_q` -> `enable_i` does make the disable work when the bus is available.

But I think If `enable_i` falls while the bus is active, `targ_bus_active` is thereafter stuck at 1, so `bus_avail_i` never returns and `buf_en_q` stays set indefinitely: the Target stays connected, out of reset, and still able to ACK, since `targ_dev[t].en` comes from `TARG_ENABLE[t]` rather than `TARG_CONTROL.EN`. It's recoverable (a reset or a re-enable clears it).

Would making `resetting` a level work, i.e. `|{~enable_i, sw_reset_i, te_recov_i}`? I think it would fix this case, keep the reset leading the buffer gating on the disable path exactly as it does on the `sw_reset` path, hold the transceiver in reset for as long as the Target is disabled (which also gives `TARG_STATUS.ACTIVE` the right value; see the review body), and make `enabling`/`enabled_q` redundant.

Agreed that deasserting the enable before a reset is the right sequence. The part that looks like a real constraint to document rather than fix is that a mid-transfer disable abandons the frame, since the Target stops driving mid-bit. My concern is only that today software has no way to guarantee disconnection while the bus is busy, which is what the RFC's pin-sharing requirement asks for.

File: hw/ip/i3c/rtl/i3c_target_rst.sv (hunk 4 above)
Comment (andreaskurth) [reply hw/ip/i3c/rtl/i3c_target_rst.sv:50]: This fix LGTM, thanks.

**Review by andreaskurth — COMMENTED** [review]
Thanks @alees24 for the rework and for the detailed replies.

I still see the need for a few changes - one substantive item (the disable path, in the existing inline thread) plus a few smaller ones.

At `i3c_core.sv:213`, `targ_status.active` still reads 1 after reset release while the Target is disabled and its input buffers are disconnected, and it reads 1 after a soft reset while `u_targ_rst` is still waiting for a Bus Available Condition. I couldn't find anything tracking it (#31307 covers debug gaps but not this field, and #31337 lists only RSTACT). Making `resetting` a level (see the inline thread) would give the right value for free; failing that, could a bullet go on #31307?

Related, the disable currently works only because `bus_avail_i` stays asserted for one more cycle after `enable_i` falls, which holds because `i3c_timers` registers `tm_resets_i`. Nothing says so, and a register stage added to `targ_rst_bus_avail` would silently stop the disable working. Moot if `resetting` becomes a level.
